File: src/main/java/de/byjoker/myjfql/user/UserServiceImpl.java
package de.byjoker.myjfql.user;

import de.byjoker.jfql.util.ID;
import de.byjoker.myjfql.core.MyJFQL;
import de.byjoker.myjfql.database.DatabaseAction;
import de.byjoker.myjfql.exception.FileException;
import de.byjoker.myjfql.util.FileFactory;
import org.json.JSONObject;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class UserServiceImpl implements UserService {

    private final FileFactory factory;
    private final List<User> users;

    public UserServiceImpl() {
        this.users = new ArrayList<>();
        this.factory = new FileFactory();
    }

    @Override
    public void createUser(User user) {
        if (getUserByName(user.getName()) != null)
            throw new FileException("User already exists!");

        if (existsUser(user.getId())) {
            user.setId(ID.generateNumber().toString());
            createUser(user);
            return;
        }

        user.setPassword(MyJFQL.getInstance().getEncryptor().encrypt(user.getPassword()));
        saveUser(user);
    }

    @Override
    public void saveUser(User user) {
        for (int i = 0; i < users.size(); i++) {
            if (users.get(i).getName().equals(user.getName())) {
                users.set(i, user);
                return;
            }
        }

        users.add(user);
    }

    @Override
    public User getUserByIdentifier(String identifier) {
        if (identifier.startsWith("#"))
            return getUser(identifier.replaceFirst("#", ""));

        return getUserByName(identifier);
    }

    @Override
    public User getUserByName(String name) {
        return users.stream().filter(user -> user.getName().equals(name)).findFirst().orElse(null);
    }

    @Override
    public User getUser(String id) {
        return users.stream().filter(user -> user.getId().equals(id)).findFirst().orElse(null);
    }

    @Override
    public boolean existsUserByIdentifier(String identifier) {
        if (identifier.startsWith("#"))
            return existsUser(identifier.replaceFirst("#", ""));

        return existsUserByName(identifier);
    }

    @Override
    public boolean existsUserByName(String name) {
        return users.stream().anyMatch(user -> user.getName().equals(name));
    }

    @Override
    public boolean existsUser(String id) {
        return users.stream().anyMatch(user -> user.getId().equals(id));
    }

    @Override
    public void deleteUser(String id) {
        users.removeIf(user -> user.getName().equals(id));
        new File("user/" + id + ".json").delete();
    }

    @Override
    public List<User> getUsers() {
        return users;
    }

    @Override
    public void loadAll() {
        loadAll(new File("user"));
    }

    @Override
    public void loadAll(File space) {
        Arrays.stream(Objects.requireNonNull(space.listFiles())).forEach(file -> {
            final JSONObject jsonUser = factory.load(file);
            final JSONObject jsonAccesses = jsonUser.getJSONObject("accesses");

            final String name = file.getName().replaceFirst(".json", "");

            if (name.contains("%") || name.contains("#") || name.contains("'")) {
                MyJFQL.getInstance().getConsole().logWarning("Database used unauthorized characters in the id!");
            } else {
                final User user = new User(name, jsonUser.getString("name"), jsonUser.getString("password"));
                user.setPreferredDatabase(jsonUser.getString("preferred"));
                user.setAccesses(jsonAccesses.keySet().stream().collect(Collectors.toMap(key -> key, key -> DatabaseAction.valueOf(jsonAccesses.getString(key)), (a, b) -> b)));

                if (!user.getName().contains("%") && !user.getName().contains("#") && !user.getName().contains("'"))
                    users.add(user);
                else
                    MyJFQL.getInstance().getConsole().logWarning("User used unauthorized characters in the name!");
            }
        });
    }

    @Override
    public void updateAll() {
        updateAll(new File("user"));
    }

    @Override
    public void updateAll(File space) {
        users.forEach(user -> {
            final File file = new File(space.getPath() + "/" + user.getId() + ".json");
            final JSONObject jsonObject = new JSONObject();
            jsonObject.put("name", user.getName());
            jsonObject.put("password", user.getPassword());
            jsonObject.put("accesses", user.getAccesses());
            jsonObject.put("preferred", user.getPreferredDatabase());
            factory.save(file, jsonObject);
        });
    }

}

File: src/main/java/de/byjoker/myjfql/user/UserServiceImpl.kt
package de.byjoker.myjfql.user

import de.byjoker.myjfql.core.MyJFQL
import de.byjoker.myjfql.database.DatabasePermissionLevel
import de.byjoker.myjfql.exception.FileException
import de.byjoker.myjfql.exception.UserException
import de.byjoker.myjfql.util.FileFactory
import de.byjoker.myjfql.util.IDGenerator.generateDigits
import de.byjoker.myjfql.util.Json
import java.io.File

class UserServiceImpl : UserService {
    private val factory: FileFactory
    private val users: MutableList<User>

    init {
        users = ArrayList()
        factory = FileFactory()
    }

    override fun createUser(user: User) {
        if (existsUserByName(user.name)) {
            throw UserException("User already exists!")
        }

        if (existsUser(user.id)) {
            user.id = generateDigits(8)
            createUser(user)
            return
        }

        user.password = MyJFQL.getInstance().encryptor.encrypt(user.password)
        saveUser(user)
    }

    override fun saveUser(user: User) {
        for (i in users.indices) if (users[i].name == user.name) {
            users[i] = user
            return
        }

        users.add(user)
    }

    override fun getUserByIdentifier(identifier: String): User? {
        return if (identifier.startsWith("#")) getUser(identifier.replaceFirst("#".toRegex(), "")) else getUserByName(
            identifier
        )
    }

    override fun getUserByName(name: String): User? {
        return users.stream().filter { user -> user.name == name }.findFirst().orElse(null)
    }

    override fun getUser(id: String): User? {
        return users.stream().filter { user -> user.id == id }.findFirst().orElse(null)
    }

    override fun existsUserByIdentifier(identifier: String): Boolean {
        return if (identifier.startsWith("#")) existsUser(
            identifier.replaceFirst(
                "#".toRegex(),
                ""
            )
        ) else existsUserByName(identifier)
    }

    override fun existsUserByName(name: String): Boolean {
        return users.stream().anyMatch { user -> user.name == name }
    }

    override fun existsUser(id: String): Boolean {
        return users.stream().anyMatch { user -> user.id == id }
    }

    override fun deleteUser(id: String) {
        users.removeIf { user -> user.id == id }
        File("user/$id.json").delete()
    }

    override fun getUsers(): List<User> {
        return users
    }

    override fun loadAll() {
        loadAll(File("user"))
    }

    override fun loadAll(backend: File) {
        if (!backend.isDirectory) {
            throw FileException("${backend.name} isn't a valid user file space!")
        }

        users.clear()

        val reserved = listOf('%', '#', '\'')

        backend.listFiles()?.iterator()?.forEach { file ->
            val node = Json.read(file)

            val user = SimpleUser(
                name = node.get("name").asText(),
                password = node.get("password").asText(),
            )

            Json.convert<Map<String, String>>(node.get("accesses")).forEach { access ->
                user.accesses[access.key] =
                    DatabasePermissionLevel.valueOf(access.value)
            }

            if (node.has("id")) {
                user.id = node.get("id").asText()
            } else {
                user.id = file.name.replace(".json", "")
            }

            if (reserved.any { char -> user.name.contains(char) || user.id.contains(char) }) {
                throw UserException("User used reserved character in id or name!")
            }

            if (existsUserByName(user.name)) {
                throw UserException("User already has been initialized!")
            }

            user.preferredDatabaseId = if (node.has("preferred")) {
                node.get("preferred").asText()
            } else {
                node.get("preferredDatabaseId").asText()
            }

            users.add(user)
        }
    }

    override fun updateAll() {
        updateAll(File("user"))
    }

    override fun updateAll(backend: File) {
        users.forEach { user -> Json.write(user, File("${backend.path}/${user.id}.json")) }
    }
}
